torch_geometric/datasets/var_misuse.py

```diff
@@ -0,0 +1,165 @@
+                    if candidate['IsCorrect']:
+                        correct_candidate_id = candidate['SymbolDummyNode']
+                    else:
+                        distractor_candidate_ids.append(
+                            candidate['SymbolDummyNode'])
+                assert correct_candidate_id is not None
+                # 0th candidate_id will always be correct
+                # candidate_node_ids = [correct_candidate_id] +
+                #                      distractor_candidate_ids
+
+                edge_indexes = []
+                edge_types = []
+
+                raw_edges = raw_sample['ContextGraph']['Edges']
+                for e_type, e_type_edges in raw_edges.items():
+                    if len(e_type_edges) > 0:
+                        e_type_idx = self.graph_edge_types_map[e_type] * 2
+                        e_type_bkwd_idx = (e_type_idx * 2) + 1
+
+                        fwd_edges = torch.tensor(
+                            e_type_edges, dtype=torch.int32).t().contiguous()
+                        fwd_edge_type = torch.tensor(
+                            [e_type_idx] * fwd_edges.size(1), dtype=torch.int8)
+                        bkwd_edges = torch.flip(fwd_edges, [0])
+                        bkwd_edge_type = torch.tensor([e_type_bkwd_idx] *
+                                                      bkwd_edges.size(1),
+                                                      dtype=torch.int8)
+
+                        edge_indexes.append(fwd_edges)
+                        edge_types.append(fwd_edge_type)
+                        edge_indexes.append(bkwd_edges)
+                        edge_types.append(bkwd_edge_type)
+
+                edge_index = torch.cat(edge_indexes, dim=1)
+                edge_type = torch.cat(edge_types, dim=-1)
+
+                num_nodes = len(raw_sample['ContextGraph']['NodeLabels'])
+                node_labels = [""] * num_nodes
+                raw_node_labels = raw_sample['ContextGraph']['NodeLabels']
+                for (node, label) in raw_node_labels.items():
+                    node_labels[int(node)] = label
+
+                data = Data(edge_index=edge_index, edge_type=edge_type,
+                            node_labels=node_labels, slot_node_id=slot_node_id,
+                            correct_candidate_id=correct_candidate_id,
+                            distractor_candidate_ids=distractor_candidate_ids)
+
+                data_list.append(data)
+
+            if self.pre_filter is not None:
+                data_list = [d for d in data_list if self.pre_filter(d)]
+
+            if self.pre_transform is not None:
+                data_list = [self.pre_transform(d) for d in data_list]
+            torch.save(data_list, processed_path)
+
+    def len(self):
+        return len(self.processed_file_names) * 100
+
+    def get(self, idx):
+        file_idx, data_idx = divmod(idx, 100)
+        window, chunk = divmod(file_idx, 10)
+        data_list = torch.load(
```

Comment:
How big is the underlying dataset on disk? Can't we use `InMemoryDataset` here?

While I like the current approach of having multiple graphs grouped together in one file, it looks like we are not really making use of this here. Currently, we are loading multiple graphs from disk, use one of them, and throw the remaining ones away.

Sadly, I currently have no good idea how to resolve this though. Any thoughts?

Reply:
@rusty1s The total sample numbers are following:

Train: 1250 * 100 = 125000
Valid: 214 * 100 = 21400
Test: 586 * 100 = 58600

Total space taken by gzipped files is around ~3.5GB for train, ~500GB for valid and ~1.5 GB for test.

The way I am processing at present,each processed pt file takes around 5 to 10MB for each chunk i.e.100 samples. So storing all train samples in memory would take around 7.5MB *1250 = ~9.375GB+

I thought that is too big for storing in memory. Most of the other in memory datasets are much smaller.

> Currently, we are loading multiple graphs from disk, use one of them, and throw the remaining ones away.

May be we can keep the pointer to the last opened chunk file open and reuse it somehow but that would only help for sequential access. I think samples would be loaded after shuffling so that would not be of much help. 100 samples or 5MB is not too big I thought to access individual sample and with SSD it likely will not be a bottleneck.

Reply:
Yes, shuffling is indeed a problem for this approach. I think 9GB would be still a reasonable amount for holding data in memory, but we can certainly think of adding a dataset interface in-between, e.g., `InMemoyChunkDataset` that effectively can re-use already loaded chunks.